mongo/collection.go

```diff
@@ -1642,34 +1570,59 @@ func (coll *Collection) Find(ctx context.Context, filter interface{},
 	return newCursorWithSession(bc, coll.bsonOpts, coll.registry, sess)
 }
 
-func newFindOptionsFromFindOneOptions(opts ...*options.FindOneOptions) []*options.FindOptions {
-	findOpts := make([]*options.FindOptions, 0, len(opts))
-	for _, opt := range opts {
-		if opt == nil {
-			continue
+func newFindOptionsFromFindOneOptions(opts ...options.Options[options.FindOneArgs]) *options.FindOptions {
```

Comment:
We can significantly simplify the logic here if we pass around `options.FindOneArgs` and `options.FindArgs` instead of passing around arrays of options. Generally, the variadic interfaces for options are syntactic sugar for the public interfaces, but we should avoid using them for internal code because they add complexity.

E.g. updated function signature for `newFindOptionsFromFindOneOptions`:
```go
func newFindArgsFromFindOneArgs(args *options.FindOneArgs) *options.FindArgs
```

Then we would create an internal, shared `find` function that has the same signature as `Find` but takes `FindArgs` instead of `...FindOptions`:
```go
func (coll *Collection) find(
	ctx context.Context,
	filter interface{},
	args *options.FindArgs,
) (cur *Cursor, err error) {
```